### ISSUE TITLE
Make it easier to copy an entire library to another library

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#3045]
https://forums.zotero.org/discussion/103877/how-to-copy-my-entire-zotero-library-to-group-zotero-library

Since you can't select multiple collections, it's pretty annoying to copy an entire library to another library. We've talked about supporting multi-collection selection, but even if we have that there should still just be a command to copy an entire library via the library context menu.

**Comment by abaevbog** [comment]
Now that we added "Copy to" and "Move to" context menu options for collections, having a "Copy to" option for the library itself seems even more relevant. It would just perform the "Copy to" operation on all its collections with the selected library as the target.
 
It does get tricky with linked collections but I suppose we would also just skip them for now

**Comment by abaevbog** [comment]
I don't suppose we want to also duplicate saved searched in that case, right? Since they can't be dragged into another library anyways. Though it might be useful for cases as described in the original forums post if you happen to have a lot of saved searches and need to share them with others.

**Comment by abaevbog** [comment]
I started looking into this, and on the first try, this just requires to copy all items from the one library to another, then to copy all collections and finally place linked items in target group into the collections. The big question is what would happen on the subsequent attempts (when linked items exist) and how would we handle conflicts between linked items and collections. There have been requests in being able to manually "sync" a collection in one group with another, so I assume this would be a more general solution to that problem.

Currently, if a collection already exists in target group, nothing happens if you try to copy it. If you try to copy an item to another group and a linked item exists, it will be added to the target collection (if any) but the actual linked item's data will not be affected. For this group copying to make sense (after the first attempt), we would need to provide a way to resolve conflicts between linked objects. One would need to be able to resolve conflicts between linked collections (select which items should end up in the collection of the target group), between items (similar to what currently happens during sync if there is a conflict), and finally handle child items of linked items.

I imagine that if collection A is being "synced" with collection B, we could display a conflict-resolution dialog (similar as to what happens during sync) where all of the above would be handled. In that dialog, we would have an itemTree with the items in the in the target collection B. It would include all current items in that collection and all items that would be added, highlighted with some neutral color. We can also highlight rows that would be removed (if the linked item was removed from source collection A) with red. +/- button-columns can be used to prevent items from being added to or removed from the target collection.

Then, we would have a mini-itemPane component that would display `itemBox` in `merge` mode to resolve potential conflicts between the data of the top-level items selected in the itemTree. If a selected item and its linked item have notes or attachments, mini-itemPane can also display options how to handle them below the `itemBox`. I think as far as child items go, we can offer to keep the children of the item that exist now, to replace them with children of linked item, or to keep both sets of attachments or notes in the item of the target collection.

This scenario I have in mind is specifically to handle passing data between two collections in different groups. For copying an entire group, we can generalize this by adding a collectionTree, so that one can go through all linked collections in the target group and resolve the conflicts in each of them.

**Comment by dstillman** [comment]
I don't think we need to provide nearly that much control here. People don't want to fiddle with that much detail, particularly for an operation that they might run repeatedly.

We can't do real syncing between libraries because we're not storing pre-change object copies or keeping a per-library-pair delete log, both of which we do for server syncing, so we can't actually know what was changed on which side — e.g., if an item's data was changed in the source library or in the target library, or if an item was removed from the source collection or added to the target collection, or if an item wasn't dragged previously or was just deleted in the target library.

So this has to be a pretty blunt operation, and probably mostly additive. So items get added to previously dragged collections, item metadata I guess gets updated, but nothing gets deleted, existing child attachments/note/annotations don't get removed (since they might have been added in the target library), subcollections on previously dragged collections missing in the source library remain, etc.

We could conceivably offer an option to do an additive copy or to completely replace the target library with the source one. I feel like people would want the latter sometimes, though it's a riskier thing to offer, and they could of course just delete everything in the target library first before copying again — but that's a little annoying for what might be a common operation, and we could do it more efficiently without causing all unchanged existing objects to be reuploaded. I do think a fairly common use case will be for someone to want to let an assistant make changes to their library and then just replace their personal library with that edited copy. We would need a pretty clear warning (e.g., using hardConfirmationDialog and typed confirmation text).

I'm not sure what any other intelligible option would be.

**Comment by abaevbog** [comment]
Having worked on this a bit, I agree that separating "copying" into two different options - purely additive and destructive - is the best way to go here. The former allows one to add newly added items/collections without being afraid that some relevant data will be overriden, while the later allowing one to do a full "sync" of a collection based on another linked collection.

I placed both options into the "Copy to" context menu of the linked collection of the selected group. Currently, the group is just disabled if the linked collection exists. But now, I am thinking to only display the linked collection and the two copy options for it. That way, I think it's very clear that the collections are linked.

This is the rough behavior of the "additive" copying I have in mind. The actual label of the menuitem may need to be tweaked - for now it's "Transfer New Items and Collections". With this, it copies all new items into the target collection (including new child items), as well as copies any unlinked subcollections as well. It does not delete or edit anything (including metadata, since we don't know which edits should take precedence).

https://github.com/user-attachments/assets/da9ed40d-2a30-44b7-9c0b-367e07e04349

And this is what the "destructive" copying can look like. Metadata and edits to child items are not affected by additive copying. This option sends the content of the collection (items and subcollections) to the trash and replaces the with the items and subcollections of another collection. That allows one to do a complete "sync" of one collection with another. Trash does serve as an insurance that if anything important is deleted, it can be recovered.

https://github.com/user-attachments/assets/0dd7b353-1efa-4a67-9caf-98e92afdb6ec

To copy a library into another library, we just perform the additive copying on all of its collections + copy any unlinked and unfiled items over.

**Comment by abaevbog** [comment]
Potential use case also addressable with the linked PR: https://forums.zotero.org/discussion/comment/502655/#Comment_502655.

If the ownership of the group is lost, you could just copy the local group elsewhere via the the context menu.
